service-api/module/Application/src/Yoti/SessionStatusService.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Application\Yoti;
+
+use Application\Fixtures\DataImportHandler;
+use Application\Model\Entity\CaseData;
+use Application\Model\Entity\CounterService;
+use Application\Yoti\Http\Exception\YotiException;
+use DateTime;
+use InvalidArgumentException;
+use Ramsey\Uuid\Uuid;
+
+class SessionStatusService
+{
+    public function __construct(
+        public readonly YotiService $yotiService,
+        public readonly DataImportHandler $dataImportHandler
+    ) {
+    }
+
+    public function getSessionStatus(CaseData $caseData): string|array|CounterService
+    {
+        if ($caseData->counterService === null) {
+            return "Error: counterService is null";
+        }
+        $currentNotificationStatus = $caseData->counterService->notificationState;
+
+        if ($currentNotificationStatus === 'first_branch_visit') {
+            return 'In Progress';
+        }
+
+        if ($currentNotificationStatus === 'session_completion' && $caseData->counterService->state !== 'COMPLETED') {
+            return $this->handleSessionCompletion($caseData);
+        }
+
+        return $caseData->counterService;
+    }
+
+    private function handleSessionCompletion(CaseData $caseData): mixed
+    {
+        $nonce = strval(Uuid::uuid4());
+        $timestamp = (new DateTime())->getTimestamp();
+        $state = '';
+        $finalResult = false;
+
+        try {
+            $response = $this->yotiService->retrieveResults($caseData->yotiSessionId, $nonce, $timestamp);
+            $state = $response['results']['state'];
+            $finalResult = $this->evaluateFinalResult($response['results']['checks']);
+
+            $this->dataImportHandler->updateCaseChildAttribute(
+                $caseData->id,
+                'counterService.state',
+                'S',
+                $state,
+            );
```

Comment:
I've just noticed that `updateCaseChildAttribute` (and `updateCaseData`) don't update the connected `CaseData` object, which means you end up with two sources of truth: what's in DynamoDB and what's in memory (i.e. if you then did `echo $caseData->counterService->state;`, it wouldn't be correct).

Probably something for a subsequent PR, but it would be good to fix that. I wonder if, instead of setting individual attributes, we could just re-put the whole object, like:
```php
$caseData->counterService->state = $state;
$this->dynamoDbClient->putItem(['TableName' => ..., 'Item' => $caseData]);
```

Reply:
Oh I missed that one, if I add `            $caseData->counterService->state = $state;
            $caseData->counterService->result = $finalResult;` this could keep them in synch for this issue? Happy to create another ticket to address other areas where the object might be out of line?

With the putItem use above I seem to remember looking into that and what I got was it would overwrite the entire Item, can look into it again

Reply:
PutItem _will_ overwrite the whole item, but that's somewhat intentional because you already have the whole item in memory (CaseData).